logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.logstreams.impl.log;
+
+import static io.camunda.zeebe.logstreams.impl.serializer.DataFrameDescriptor.FRAME_ALIGNMENT;
+
+import io.camunda.zeebe.logstreams.impl.serializer.DataFrameDescriptor;
+import io.camunda.zeebe.logstreams.log.LogAppendEntry;
+import io.camunda.zeebe.logstreams.log.LogStreamBatchWriter;
+import io.camunda.zeebe.scheduler.ActorCondition;
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The sequencer is a multiple-producer, single-consumer queue of {@link LogAppendEntry}. It buffers
+ * a fixed amount of entries and rejects writes when the queue is full. The consumer may read at its
+ * own pace by repeatedly calling {@link Sequencer#tryRead()} or register for notifications when new
+ * entries are written by calling {@link Sequencer#registerConsumer(ActorCondition)}.
+ *
+ * <p>The sequencer assigns all entries a position and makes that position available to its
+ * consumer. The sequencer does not copy or serialize entries, it only keeps a reference to them
+ * until they are handed off to the consumer.
+ */
+public final class Sequencer implements LogStreamBatchWriter, Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(Sequencer.class);
+  private final int partitionId;
+  private final int maxFragmentSize;
+
+  private volatile long position;
+  private volatile boolean isClosed = false;
+  private volatile ActorCondition consumer;
+  private final Queue<SequencedBatch> queue = new ArrayBlockingQueue<>(128);
+  private final ReentrantLock lock = new ReentrantLock();
+  private final SequencerMetrics metrics;
+
+  public Sequencer(final int partitionId, final long initialPosition, final int maxFragmentSize) {
+    LOG.trace("Starting new sequencer at position {}", initialPosition);
+    this.position = initialPosition;
+    this.partitionId = partitionId;
+    this.maxFragmentSize = maxFragmentSize;
+    this.metrics = new SequencerMetrics(partitionId);
+  }
+
+  /**
+   * @param eventCount the potential event count we want to check
+   * @param batchSize the potential batch Size (in bytes) we want to check
+   * @return True if the serialized batch would fit within {@link Sequencer#maxFragmentSize}, false
+   *     otherwise.
+   */
+  @Override
+  public boolean canWriteEvents(final int eventCount, final int batchSize) {
+    final int framedMessageLength =
```

Comment:
:thought_balloon: This logic is kind of tied to the serializer, in a way. I wonder if we could somehow centralize this? It's not a big deal though if we can't.